### PR TITLE
Quasar 17.20.0 plat 23554

### PR DESCRIPTION
**Pull request by danbar0kal**

### CODE REVIEW

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	const VIRTUAL_EVENT_ID = 'virtualEventId';
+	const VERSION = 'version';
+	const ANSWER_IDS = 'answerIds';
+	const SCORE = 'score';
+	const NUM_OF_CORRECT_ANSWERS = 'numOfCorrectAnswers';
+	const NUM_OF_QUESTIONS = 'numOfQuestions';
+	const NUM_OF_RELEVANT_QUESTIONS = 'numOfRelevnatQuestions';
+	const CALCULATED_SCORE = 'calculatedScore';
+	
+	const MAP_NAME = 'internal_analytics_host';
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:22]: this is a param name, also rename the ver name

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	const HEADER_HOST = 'Host';
+	const HEADER_IDENTIFIER = 'X-Forwarded-For';
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:24]: rename header var names

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+		{
+			return true;
+		}
+		return false;
+	}
+	
+	protected static function wasStatusChanged(BaseObject $object, array $modifiedColumns)
+	{
+		return ($object instanceof QuizUserEntry)
+			&& in_array('user_entry.STATUS', $modifiedColumns)
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:98]: use enum UserEntryPeer::STATUS
and QuizPlugin::getCoreValue('UserEntryStatus', QuizUserEntryStatus::QUIZ_SUBMITTED)

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	
+	/**
+	 * @param BaseObject $object
+	 * @param array $modifiedColumns
+	 * @return bool
+	 */
+	public function objectChanged (BaseObject $object, array $modifiedColumns)
+	{
+		$quizEventContent = array();
+		if (self::wasStatusChanged($object, $modifiedColumns))
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:36]: if(!wasStatus...)
return
then send logic

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+			$quizEventContent = self::generateQuizEventContent($object);
+		}
+		$uri = '/api_v3/index.php?' . http_build_query($quizEventContent, '', '&');
+		
+		self::sendBeacon($uri);
+	}
+	
+	protected static function sendBeacon (string $uri)
+	{
+		$statsHost = explode(':', kConf::get(self::MAP_NAME));
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:47]: use internal_analytics_host const and pass it to the function

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+		}
+		$out = "GET {$uri} HTTP/1.1\r\n";
+		foreach($headers as $header => $value)
+		{
+			$out .= "$header: $value\r\n";
+		}
+		$out .= "\r\n";
+		
+		self::sendRequest($uri, $host, $port, $out);
+
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:65]: remove new line

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+		if (self::wasStatusChanged($object, $modifiedColumns))
+		{
+			$quizEventContent = self::generateQuizEventContent($object);
+		}
+		$uri = '/api_v3/index.php?' . http_build_query($quizEventContent, '', '&');
+		
+		self::sendBeacon($uri);
+	}
+	
+	protected static function sendBeacon (string $uri)
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:45]: move the duplicated code to function which will be used for both of the deacon existing types

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+		return ($object instanceof QuizUserEntry)
+			&& in_array('user_entry.STATUS', $modifiedColumns)
+			&& $object->getStatus() == 10060;
+	}
+	
+	protected static function generateQuizEventContent($quizUserEntry)
+	{
+		/* @var $quizUserEntry QuizUserEntry */
+		return array(
+			self::EVENT_TYPE => $quizUserEntry->getType(),
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:106]: we will have a fixed event type

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+			self::KUSER_ID => $quizUserEntry->getkuser()->getEmail(),
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:107]: we want the kuser id number (getId)

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+			self::PARTNER_ID => $quizUserEntry->getPartnerId(),
+			self::ENTRY_ID => $quizUserEntry->getEntryId(),
+			self::EVENT_TIME => date('Y-m-d H:i:s'),//"2022-01-19T13:32:10Z" current time of this event being issued
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:110]: can remove event time and VIRTUAL_EVENT_ID field

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+			self::VIRTUAL_EVENT_ID => "Unknown",
+			self::VERSION => $quizUserEntry->getVersion(),
+			self::ANSWER_IDS => $quizUserEntry->getAnswerIds(),
+			self::SCORE => $quizUserEntry->getScore(),//check
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:114]: remove comment

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+	public function shouldConsumeChangedEvent (BaseObject $object, array $modifiedColumns)
+	{
+		if (self::wasStatusChanged($object, $modifiedColumns))
+		{
+			return true;
+		}
+		return false;
+	}
+	
+	protected static function wasStatusChanged(BaseObject $object, array $modifiedColumns)
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:95]: move save in user entry to make sure this event includes all the quiz properties

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+	 * @param array $modifiedColumns
+	 * @return bool
+	 */
+	public function objectChanged (BaseObject $object, array $modifiedColumns)
+	{
+		if (!self::wasStatusChanged($object, $modifiedColumns))
+		{
+			return false;
+		}
+		$quizEventContent = self::getQuizEventContent($object);
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:33]: move this under the if, there is no point getting the content if we are not going to send the beacon

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+			$statsHost = explode(':', kConf::get(self::INTERNAL_ANALYTICS_HOST));
+			self::sendBeacon(
+				$quizEventContent,
+				$statsHost[0],
+				isset($statsHost[1]) ? $statsHost[1] : 80);
+		}
+		return true;
+	}
+	
+	public static function sendBeacon ($quizEventContent, $host, $port)
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:45]: dont think we need this function just as a proxy to another function

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+		return ($object instanceof QuizUserEntry)
+			&& in_array(UserEntryPeer::STATUS, $modifiedColumns)
+			&& $object->getStatus() == QuizPlugin::getCoreValue('UserEntryStatus', QuizUserEntryStatus::QUIZ_SUBMITTED);
+	}
+	
+	protected static function getQuizEventContent($quizUserEntry)
+	{
+		/* @var $quizUserEntry QuizUserEntry */
+		return array(
+			self::EVENT_TYPE => self::QUIZ_EVENT_TYPE,
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:77]: need to remember to change it to the relevant id

File: plugins/cue_points/quiz/lib/kQuizManager.php [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+			self::KUSER_ID => $quizUserEntry->getkuser()->getId(),
```
Comment (inbal-ben-david) [inline plugins/cue_points/quiz/lib/kQuizManager.php:78]: separate the calls, if we wont get kuser from getKuser you will end up with null exception
add the kuserId only if kuser returned

File: alpha/apps/kaltura/lib/requestUtils.class.php [hunk 17]
```diff
@@ -329,4 +333,36 @@ public static function matchIpCountry ( $ip_country_list_str , &$current_country
 		$current_country = self::getIpCountry() ;
 		return ( in_array ( $current_country , $ip_country_list ) );
 	}
+	
+	public static function buildKavaRequest($quizEventContent, $host)
```
Comment (inbal-ben-david) [inline alpha/apps/kaltura/lib/requestUtils.class.php:337]: rename to sendAnalyticsBeacon and rename quizEventContent to something generic

File: alpha/apps/kaltura/lib/requestUtils.class.php [hunk 18]
```diff
@@ -329,4 +333,36 @@ public static function matchIpCountry ( $ip_country_list_str , &$current_country
...
+		foreach($headers as $header => $value)
+		{
+			$out .= "$header: $value\r\n";
+		}
+		$out .= "\r\n";
+		
+		return $out;
+	}
+	
+	public static function sendKavaRequest($host, $port, $out)
```
Comment (inbal-ben-david) [inline alpha/apps/kaltura/lib/requestUtils.class.php:357]: can be combined with the function above as it just sends the beacon

File: plugins/cue_points/quiz/lib/kQuizManager.php (hunk 16 above)
Comment (danbar0kal) [reply plugins/cue_points/quiz/lib/kQuizManager.php:78]: what should be the value if kuser is null?

**Review by inbal-ben-david — APPROVED** [review]

**Pull request merged**